src/tdycore.c

```diff
@@ -1333,11 +1371,106 @@ PetscErrorCode TDySetDtimeForSNESSolver(TDy tdy, PetscReal dtime) {
+/// e.g. TDyGetSaturation to store the saturation field in such a vector.
+/// @param [in] tdy A TDy object
+PetscErrorCode TDyUpdateDiagnostics(TDy tdy) {
   PetscErrorCode ierr;
+  PetscFunctionBegin;
 
+  MPI_Comm comm;
+  ierr = PetscObjectGetComm((PetscObject)tdy, &comm); CHKERRQ(ierr);
+
+  if ((tdy->setup_flags & TDySetupFinished) == 0) {
+    SETERRQ(comm,PETSC_ERR_USER,"You must call TDyUpdateDiagnostics after TDySetup()");
+  } else if (!tdy->ops->update_diagnostics) {
+    SETERRQ(comm,PETSC_ERR_USER,"TDyUpdateDiagnostics is not supported by this implementation.");
+  }
+
+  // Update the diagnostic fields.
+  ierr = tdy->ops->update_diagnostics(tdy->context, tdy->diag_dm, tdy->diag_vec);
+  CHKERRQ(ierr);
+
+  PetscFunctionReturn(0);
+}
+
+/// Creates a Vec that can store a cell-centered scalar diagnostic field such as
+/// the saturation or liquid mass.
+/// @param [in] tdy A TDy object
+/// @param [out] diag_vec A Vec that can store a diagnostic field.
+PetscErrorCode TDyCreateDiagnosticVector(TDy tdy, Vec *diag_vec) {
+  PetscErrorCode ierr;
   PetscFunctionBegin;
-  ierr = TDyNaturalToGlobal(tdy,initial,tdy->solution); CHKERRQ(ierr);
+
+  MPI_Comm comm;
+  ierr = PetscObjectGetComm((PetscObject)tdy, &comm); CHKERRQ(ierr);
+
+  if ((tdy->setup_flags & TDySetupFinished) == 0) {
+    SETERRQ(comm,PETSC_ERR_USER,"You must call TDyCreateDiagnosticVector after TDySetup()");
+  } else if (!tdy->ops->update_diagnostics) {
+    SETERRQ(comm,PETSC_ERR_USER,"Diagnostic fields are not supported by this implementation.");
+  }
+
+  // Create a cell-centered scalar field vector.
+  PetscInt c_start, c_end;
+  ierr = DMPlexGetHeightStratum(tdy->diag_dm, 0, &c_start, &c_end); CHKERRQ(ierr);
+  ierr = VecCreateMPI(comm, c_end - c_start, PETSC_DECIDE, diag_vec); CHKERRQ(ierr);
+
+  PetscFunctionReturn(0);
+}
+
+static PetscErrorCode ExtractDiagnosticField(TDy tdy, PetscInt index, Vec vec) {
+  PetscErrorCode ierr;
+  PetscFunctionBegin;
+
+  MPI_Comm comm;
+  ierr = PetscObjectGetComm((PetscObject)tdy, &comm); CHKERRQ(ierr);
+
+  if ((tdy->setup_flags & TDySetupFinished) == 0) {
+    SETERRQ(comm,PETSC_ERR_USER,"Diagnostic fields cannot be extracted before TDySetup()");
+  } else if (!tdy->ops->update_diagnostics) {
+    SETERRQ(comm,PETSC_ERR_USER,"Diagnostic fields are not supported by this implementation.");
+  }
+
+  // Extract the field.
+  ierr = VecStrideGather(tdy->diag_vec, index, vec, INSERT_VALUES);
+  CHKERRQ(ierr);
+
+  PetscFunctionReturn(0);
+}
+
+/// Given a Vec created with TDyCreateDiagnosticVector, populates that vector
+/// with the saturation values for each cell in the grid.
+/// @param [in] tdy A TDy object
+/// @param [out] sat_vec The Vec that stores the cell-centered saturation field
+PetscErrorCode TDyGetSaturation(TDy tdy, Vec sat_vec) {
+  PetscErrorCode ierr;
+  PetscFunctionBegin;
+  ierr = ExtractDiagnosticField(tdy, 0, sat_vec); CHKERRQ(ierr);
```

Comment:
Regarding the `index` (`= 0`), can we have a variable name that is obvious? Not sure if we can have a global variable e.g. `DIAG_SATURATION`. Another option would be to pass the field name (`"Saturation"`) and determine the offset for that field using `PetscSectionGetFieldOffset`.

Reply:
I changed the name to "LiquidSaturation" and `TDyGetLiquidSaturation` to match the liquid mass stuff, and introduced these diagnostic field index macros:

* `DIAG_LIQUID_SATURATION` (0)
* `DIAG_LIQUID_MASS` (1)

Good suggestion.